src/KingSystem/Sound/sndMusicInfo.h

```diff
@@ -0,0 +1,130 @@
+#pragma once
+
+#include <agl/Utils/aglParameter.h>
+#include <agl/Utils/aglParameterIO.h>
+#include <agl/Utils/aglParameterList.h>
+#include <agl/Utils/aglParameterObj.h>
+#include <container/seadBuffer.h>
+#include <resource/seadResource.h>
+#include "KingSystem/Utils/Types.h"
+
+namespace ksys::snd {
+
+class MusicInfo : public agl::utl::IParameterIO {
+public:
+    template <typename T>
+    struct MinMax : public agl::utl::IParameterObj {
+        inline MinMax(T min, T max, const sead::SafeString& name, const sead::SafeString& label,
+                      const sead::SafeString& meta, agl::utl::IParameterList* parent)
+            : mLabel(label), mMeta(meta), mMin(min, "Min", "最小", mMeta, this),
+              mMax(max, "Max", "最大", mMeta, this) {
+            parent->addObj(this, name);
+        }
+
+        sead::FixedSafeString<64> mLabel;
+        sead::FixedSafeString<32> mMeta;
+        agl::utl::Parameter<T> mMin;
+        agl::utl::Parameter<T> mMax;
+    };
+
+    MusicInfo() : IParameterIO("mscinfo", 0) {}
+    // TODO: 0x7101015628, requires `Sound` and `snd::ResourceMgr`
+    sead::DirectResource* loadMusic();
+
+protected:
+    sead::SafeString mBgmTypeName;
+};
+
+// TODO rename this to EnvBgm
+class MusicParam : public MusicInfo {
+public:
+    SEAD_ENUM(EnvTimeEnum,EnvDayBgm,EnvNightBgm)
+
+    struct ParamPerType : public agl::utl::IParameterObj {
+        ParamPerType();
+
+        agl::utl::Parameter<unsigned int> mBPM;
+        agl::utl::Parameter<unsigned int> mPart0FirstWaitCount;
+        agl::utl::Parameter<unsigned int> mPart1FirstWaitCount;
+        agl::utl::Parameter<unsigned int> mPart2FirstWaitCount;
+        agl::utl::Parameter<unsigned int> mPart3FirstWaitCount;
+    };
+    KSYS_CHECK_SIZE_NX150(ParamPerType, 0xD0);
+
+    MusicParam();
+    void init(sead::Heap* heap);
+
+private:
+    agl::utl::IParameterList mHeader;
+    agl::utl::IParameterObj mData;
+    sead::Buffer<MusicParam::ParamPerType> mParamPerTypes;
```

Comment:
We might want to make these public or add getters so that the parameters can actually be used.

Reply:
Which is better?